raft: enforce invariant matchCommit <= sentCommit

iskettaneh opened the pull request:
This commit ensures that the matchCommit is always less than or equal to the sentCommit. This avoids sending an extra empty MsgApp. Moreover, it adds SentCommit and MatchCommit to Progress.String() to be explicit in the tests.

Epic: None

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/129943)
<!-- Reviewable:end -->

pav-kv commented on pkg/raft/tracker/progress_test.go:
```diff
@@ -36,7 +36,8 @@ func TestProgressString(t *testing.T) {
 		Inflights:          ins,
 	}
-	const exp = `StateSnapshot match=1 next=2 learner paused pendingSnap=123 inactive inflight=1[full]`
+	const exp = "StateSnapshot match=1 next=2 sentCommit=0 matchCommit=0 learner paused " +
```
nit: set some non-zero `sent/matchCommit`?

pav-kv commented on pkg/raft/testdata/snapshot_succeed_via_app_resp_behind.txt:
```diff
@@ -165,6 +165,6 @@ stabilize 1
-1: StateReplicate match=12 next=13
-2: StateReplicate match=12 next=13
-3: StateReplicate match=11 next=13 inflight=1
+1: StateReplicate match=12 next=13 sentCommit=0 matchCommit=11
```
Looks like we should have an invariant `sentCommit >= matchCommit`, similarly to `Next > Match`.

Otherwise, the logic in `ShouldSendMsgApp` can send an unnecessary message in some cases. For example in this `sentCommit=0 matchCommit=11` state we will send a message even if `commit == 11`?

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129943)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv)

___
*[`pkg/raft/testdata/snapshot_succeed_via_app_resp_behind.txt` line 168 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/129943#gh-1742455177:-O5t4b5YEL7D1zuvjKyi:b2pjrug) ([raw file](https://github.com/cockroachdb/cockroach/blob/2405a9956e3ca0f17a42a0c08ab394d743683fce/pkg/raft/testdata/snapshot_succeed_via_app_resp_behind.txt#L168)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Looks like we should have an invariant `sentCommit >= matchCommit`, similarly to `Next > Match`.

Otherwise, the logic in `ShouldSendMsgApp` can send an unnecessary message in some cases. For example in this `sentCommit=0 matchCommit=11` state we will send a message even if `commit == 11`?
</blockquote></details>

Good point! I have just addressed that


<!-- Sent from Reviewable.io -->

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129943)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @iskettaneh)

___
*[`pkg/raft/tracker/progress.go` line 57 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/129943#-O5wjul6ADCObTs6hjpr:-O5wjul6ADCObTs6hjps:b-p0480) ([raw file](https://github.com/cockroachdb/cockroach/blob/8292ee6d6e210e0af15034262484892f9e029e8d/pkg/raft/tracker/progress.go#L57)):*
> ```Go
> 	//
> 	// In StateSnapshot, SentCommit == PendingSnapshot == Next-1.
> 	// Invariant: SentCommit >= MatchCommit
> ```

Are there places other than `MaybeUpdateMatch` commit where this invariant needs to be uphold? I'm worried that some methods like `BecomeProbe` or `UpdateSentCommit` might accidentally break it. Can we add some safety guardrails (e.g. see one suggestion below)?
___
*[`pkg/raft/tracker/progress.go` line 214 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/129943#-O5wkYAKE7jWvcncPEOw:-O5wkYAKE7jWvcncPEOx:bsqvqwu) ([raw file](https://github.com/cockroachdb/cockroach/blob/8292ee6d6e210e0af15034262484892f9e029e8d/pkg/raft/tracker/progress.go#L214)):*
> ```Go
> // UpdateSentCommit updates the SentCommit.
> func (pr *Progress) UpdateSentCommit(commit uint64) {
> 	pr.SentCommit = commit
> ```

How about

```golang
if commit >= pr.SentCommit {
	pr.SentCommit = commit
}
```

This will guard the `SentCommit >= MatchCommit` invariant. Also, it will allow the caller to not check `if commit != 0` and call this unconditionally.
___
*[`pkg/raft/tracker/progress.go` line 232 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/129943#-O5wjL1u5QoIUIiPFw83:-O5wjL1u5QoIUIiPFw84:b-85arys) ([raw file](https://github.com/cockroachdb/cockroach/blob/8292ee6d6e210e0af15034262484892f9e029e8d/pkg/raft/tracker/progress.go#L232)):*
> ```Go
> // larger than the previous match commit.
> func (pr *Progress) MaybeUpdateMatchCommit(commit uint64) {
> 	pr.MatchCommit = max(commit, pr.MatchCommit)
> ```

nit:
```golang
if commit > pr.MatchCommit {
	pr.MatchCommit = commit
	p.SentCommit = max(pr.SentCommit, commit)
}
```

if `commit <= pr.MatchCommit`, we don't need to update `SentCommit`, since the invariant is true before this call, so it will stay true if we do nothing.
___
*[`pkg/raft/testdata/snapshot_succeed_via_app_resp_behind.txt` line 168 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/129943#gh-1742455177:-O5wj6ev6SinMcMSNfh5:b-jyep35) ([raw file](https://github.com/cockroachdb/cockroach/blob/2405a9956e3ca0f17a42a0c08ab394d743683fce/pkg/raft/testdata/snapshot_succeed_via_app_resp_behind.txt#L168)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

Good point! I have just addressed that
</blockquote></details>

Nice. Do you mind splitting this fix out into a separate commit in this PR, so that we can review it in isolation?


<!-- Sent from Reviewable.io -->

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129943)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @iskettaneh)

___
*[`pkg/raft/tracker/progress.go` line 214 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/129943#-O5wkYAKE7jWvcncPEOw:-O5wl1NW0IbJ2ZLhRzcs:bopvrsu) ([raw file](https://github.com/cockroachdb/cockroach/blob/8292ee6d6e210e0af15034262484892f9e029e8d/pkg/raft/tracker/progress.go#L214)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

How about

```golang
if commit >= pr.SentCommit {
	pr.SentCommit = commit
}
```

This will guard the `SentCommit >= MatchCommit` invariant. Also, it will allow the caller to not check `if commit != 0` and call this unconditionally.
</blockquote></details>

typo: `if commit > pr.SentCommit`


<!-- Sent from Reviewable.io -->

iskettaneh committed:
raft: enforce invariant matchCommit <= sentCommit

This commit ensures that the matchCommit is always less than or equal to
the sentCommit. This avoids sending an extra empty MsgApp.

Epic: None

Release note: None

iskettaneh committed:
raft: add SentCommit and MatchCommit to Progress.String()

This commit does the following:

1) Adds SentCommit & MatchCommit to Progress.String(). This helps assert
that things are correct in tests.

2) Resets `MatchCommit` to 0 to match what we do with the `Match` field.

Epic: None

Release note: None

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129943)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv)

___
*[`pkg/raft/tracker/progress.go` line 57 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/129943#-O5wjul6ADCObTs6hjpr:-O5x8Zwg4_0JZkaIhh5C:bv4odgv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8292ee6d6e210e0af15034262484892f9e029e8d/pkg/raft/tracker/progress.go#L57)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Are there places other than `MaybeUpdateMatch` commit where this invariant needs to be uphold? I'm worried that some methods like `BecomeProbe` or `UpdateSentCommit` might accidentally break it. Can we add some safety guardrails (e.g. see one suggestion below)?
</blockquote></details>

The MatchCommit is only updated in one place, so it's easy to argue that the match commit updates won't break the invariant.

The SentCommit is updated in multiple places, and I had to think about every one of them:
1) BecomeProbe: pr.SentCommit = min(pr.SentCommit, pr.Next-1): The MatchCommit should always be less than or equal pr.Next-1
2) BecomeSnapshot: pr.SentCommit basically equals pr.Next-1. The MatchCommit should always be less than or equal pr.Next-1
3) MaybeDecrTo: pr.SentCommit = min(pr.SentCommit, pr.Next-1): Same argument as above
4) UpdateSentCommit: Your suggested guardrail should help in this case.
___
*[`pkg/raft/testdata/snapshot_succeed_via_app_resp_behind.txt` line 168 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/129943#gh-1742455177:-O5xHBLIAjaM2MOByen3:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/2405a9956e3ca0f17a42a0c08ab394d743683fce/pkg/raft/testdata/snapshot_succeed_via_app_resp_behind.txt#L168)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Nice. Do you mind splitting this fix out into a separate commit in this PR, so that we can review it in isolation?
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

pav-kv commented on pkg/raft/testdata/msg_app_commit_index.txt:
```diff
@@ -189,3 +189,9 @@ stabilize 1 2 3
+
+status 1
+----
+1: StateReplicate match=13 next=14 sentCommit=11 matchCommit=11
```
I wonder why the commit haven't converged here, after `stabilize`. Is this because we're the leader, and we don't keep these indices up to date on the leader? Might be worth fixing that for consistency (in another commit, as always :))?

This resembles the inconsistency that we had for `Match/Next`, fixed in #126902.

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129943)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv)

___
*[`pkg/raft/testdata/msg_app_commit_index.txt` line 195 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/129943#gh-1744025389:-O5xhBzc4LgEF3WhdxjO:b-isotgy) ([raw file](https://github.com/cockroachdb/cockroach/blob/69eaf334fdf9b5f377c94225a8904babcf3ae243/pkg/raft/testdata/msg_app_commit_index.txt#L195)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

I wonder why the commit haven't converged here, after `stabilize`. Is this because we're the leader, and we don't keep these indices up to date on the leader? Might be worth fixing that for consistency (in another commit, as always :))?

This resembles the inconsistency that we had for `Match/Next`, fixed in #126902.
</blockquote></details>

Correct. we don't track/care about them on the leader. But yeah, I will follow the same PR that you suggested to fix the inconsistency.


<!-- Sent from Reviewable.io -->

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129943)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv)

___
*[`pkg/raft/raft.go` line 728 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129943#gh-1744239190:-O5yL-qy8IgzPB6QrSHf:b-zd5r5m) ([raw file](https://github.com/cockroachdb/cockroach/blob/9029fca91fe15bf6bb6f61a70628f05fd65b0096/pkg/raft/raft.go#L728)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

It's ok to update `SentCommit` here, but, for consistency, we need to condition `MatchCommit` update to the fact that the commit index is persisted.

The leader does a self-directed `MsgAppResp` (which eventually updates the leader's `Match`). Can we update `MatchCommit` there too? We do it in `MsgAppResp` handler in `stepLeader`, so I'm surprised `MatchCommit` wasn't up-to-date.

Also, should we move `SentCommit` update to `raft.maybeCommit` method where the leader first learns about the index advancement?
</blockquote></details>

The self-directed MsgAppResp is only sent when the leader appends some entries to the log. That's why the MatchCommit was stale in some tests because sometimes the commit index is advanced without receiving a MsgProp.

The other code path is what was modified in this commit. When the leader learns about a commit index advancement, it sends a MsgApp to all followers (but not itself). Do you suggest that we send a MsgAppResp to ourselves in that case?

Regarding moving `SentCommit` to `raft.maybeCommit`. I noticed that in some cases (where there has been recently a leader change). We might call maybeCommit even if the progress is nil. That's why I added it inside the broadcast MsgApp because we won't sent a MsgApp in the case above (since there are no nodes to iterate over).

Do you think I should keep the code as it is right now, or shall I add a check that pr is not nil in the function `MaybeUpdateSentCommit()`?


<!-- Sent from Reviewable.io -->

iskettaneh commented:
*[`pkg/raft/raft.go` line 728 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129943#gh-1744239190:-O61XnOOD7nMWG3xzuNT:bc1ap13) ([raw file](https://github.com/cockroachdb/cockroach/blob/9029fca91fe15bf6bb6f61a70628f05fd65b0096/pkg/raft/raft.go#L728)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

The self-directed MsgAppResp is only sent when the leader appends some entries to the log. That's why the MatchCommit was stale in some tests because sometimes the commit index is advanced without receiving a MsgProp.

The other code path is what was modified in this commit. When the leader learns about a commit index advancement, it sends a MsgApp to all followers (but not itself). Do you suggest that we send a MsgAppResp to ourselves in that case?

Regarding moving `SentCommit` to `raft.maybeCommit`. I noticed that in some cases (where there has been recently a leader change). We might call maybeCommit even if the progress is nil. That's why I added it inside the broadcast MsgApp because we won't sent a MsgApp in the case above (since there are no nodes to iterate over).

Do you think I should keep the code as it is right now, or shall I add a check that pr is not nil in the function `MaybeUpdateSentCommit()`?
</blockquote></details>

For the record, I think it's safe NOT to persist the leader's MatchCommit because it's not used for anything. My worry was if we do a leader transfer, the new leader might receive the wrong MatchCommit for the old leader, and that it might cause an issue. However, the new leader will try to send a MsgApp using the new term. Therefore, it will learn all the followers' (and the old leader's) MatchCommit.


<!-- Sent from Reviewable.io -->

pav-kv reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129943)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @iskettaneh)

___
*[`pkg/raft/tracker/progress.go` line 57 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/129943#-O5wjul6ADCObTs6hjpr:-O67yobk78D2Ry_FBRxK:baey3lf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8292ee6d6e210e0af15034262484892f9e029e8d/pkg/raft/tracker/progress.go#L57)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

The MatchCommit is only updated in one place, so it's easy to argue that the match commit updates won't break the invariant.

The SentCommit is updated in multiple places, and I had to think about every one of them:
1) BecomeProbe: pr.SentCommit = min(pr.SentCommit, pr.Next-1): The MatchCommit should always be less than or equal pr.Next-1
2) BecomeSnapshot: pr.SentCommit basically equals pr.Next-1. The MatchCommit should always be less than or equal pr.Next-1
3) MaybeDecrTo: pr.SentCommit = min(pr.SentCommit, pr.Next-1): Same argument as above
4) UpdateSentCommit: Your suggested guardrail should help in this case.
</blockquote></details>

Is `MatchCommit < Next` an invariant that we maintain? I think it can be violated when we call `MaybeUpdateMatchCommit` before `MaybeUpdate`, but it might be the case that after both calls this invariant is restored. I can see the logic behind it, but I don't think it's rigorously true.

Maybe it's fine to say that `SentCommit >= MatchCommit` is a best-effort invariant? We don't strongly depend on it for correctness. The most important bit is the correctness of `MatchCommit` tracking, and if `SentCommit` lags for some reason, it will cause an extra message but eventually it will converge.


<!-- Sent from Reviewable.io -->

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129943)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv)

___
*[`pkg/raft/tracker/progress.go` line 57 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/129943#-O5wjul6ADCObTs6hjpr:-O68BILO3Q9SE7nT3KuU:bst9a47) ([raw file](https://github.com/cockroachdb/cockroach/blob/8292ee6d6e210e0af15034262484892f9e029e8d/pkg/raft/tracker/progress.go#L57)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Is `MatchCommit < Next` an invariant that we maintain? I think it can be violated when we call `MaybeUpdateMatchCommit` before `MaybeUpdate`, but it might be the case that after both calls this invariant is restored. I can see the logic behind it, but I don't think it's rigorously true.

Maybe it's fine to say that `SentCommit >= MatchCommit` is a best-effort invariant? We don't strongly depend on it for correctness. The most important bit is the correctness of `MatchCommit` tracking, and if `SentCommit` lags for some reason, it will cause an extra message but eventually it will converge.
</blockquote></details>

I changed the comment to reflect that.


<!-- Sent from Reviewable.io -->

iskettaneh commented:
bors r+
